Mathlib/Tactic/Linter/DeprecateModule.lean

```diff
@@ -0,0 +1,138 @@
+* a module `Name` that has been deprecated and
+* an array of `Name`s of modules that should be imported instead
+
+to the environment.
+-/
+initialize deprecateModuleExt :
+    SimplePersistentEnvExtension (Name × Array Name) (Std.HashSet (Name × Array Name)) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := fun as => as.foldl Std.HashSet.insertMany {}
+    addEntryFn := .insert
+  }
+
+/--
+`addModuleDeprecation` adds to the `deprecateModuleExt` extension the pair consisting of the
+current module name and the array of its direct imports.
+
+It ignores `Init`, since this is a special module that is expected to be imported by all files.
+-/
+def addModuleDeprecation {m : Type → Type} [Monad m] [MonadEnv m] [MonadQuotation m] : m Unit := do
+  modifyEnv (deprecateModuleExt.addEntry ·
+    (← getMainModule, (← getEnv).imports.filterMap fun i =>
+      if i.module == `Init then none else i.module))
+
+/--
+`deprecate_module since yyyy-mm-dd` deprecates the current module `A` in favour of its imports.
+This means that any file that imports `A` will get a notification on the `import A` line suggesting
+to instead import the *direct imports* of `A`.
+-/
+elab (name := deprecate_modules)
+    "deprecate_module " &"since" yyyy:num "-" mm:num "-" dd:num: command => do
+  if yyyy.getNat < 2025 then
+    throwErrorAt yyyy "The year should be at least 2025!"
+  if mm.getNat == 0 || 12 < mm.getNat || mm.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt mm "The month should be of the form 01, 02,..., 12!"
+  if dd.getNat == 0 || 31 < dd.getNat || dd.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt dd "The day should be of the form 01, 02,..., 31!"
+  addModuleDeprecation
+  -- disable the linter, so that it does not complain in the file with the deprecation
+  elabCommand (← `(set_option linter.deprecateModule false))
+
+/-- A utility function to show the pairings `(deprecatedModule, #[preferredModules])`. -/
+elab "show_deprecated_modules" : command => do
+  let directImports := deprecateModuleExt.getState (← getEnv)
+  logInfo <| "\n".intercalate <|
+    directImports.fold (init := ["Deprecated modules\n"]) fun nms (i, deps) =>
+      nms ++ [s!"'{i}' deprecates to\n{deps}\n"]
+
+namespace DeprecateModule
+
+@[inherit_doc Mathlib.Linter.linter.deprecateModule]
+def deprecateModuleLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.deprecateModule (← getOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  let mainModule ← getMainModule
+  unless Linter.getLinterValue linter.style.header (← getOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  -- The imports of `Mathlib.lean` and `Mathlib.Tactic` are not required to be non-deprecated,
+  -- since these files are expected to import all modules satisfying certain path-restrictions.
```

Comment:
How about:
```suggestion
  -- `Mathlib.lean` and `Mathlib/Tactic.Lean` are allowed to import deprecated files,
  -- since they (are expected to) import all modules in certain directories.